### PR TITLE
python 3.12 upgrade

### PR DESCRIPTION
**Pull request by chris48s**
And so the "upgrade all the things" project rumbles on.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -11,7 +11,7 @@ project_urls =
     Tracker = https://github.com/DemocracyClub/dc_logging/issues
 
 [options]
-python_requires = >=3.8
+python_requires = >=3.11
```
Comment (chris48s) [inline setup.cfg:14]: It looks like there are 2 places we care about using this package:

1. In our applications (which are all now running on python 3.12, across the board)
2. In AWS CloudShell.

TIL AWS CloudShell. TI also L that the most recent python version available in the Amazon Linux 2023 repository is still python 3.11. Obviously there are other ways to get python on there (pyenv, uv, etc) but it seems like the lowest friction thing to do here is say we are compatible with >=3.11 and leave the instructions in https://github.com/DemocracyClub/dc_logging/blob/main/dc_logging_legacy_import/README.md#running-from-aws-cloudshell as they are for now.

I think we can live with the test suite only being run against 3.12 though. Does that seem right to you?

**Comment by chris48s** [comment]
Suspect this one might end up sitting around open for a bit, so some notes:

I tried deploying this to the dev account. When deployed to dev I can log a message and the log pops up in the S3 bucket as expected. (I've now rolled those commits back). The code works fine on python 3.12. It is the mocking that is causing the test failure.

I managed to track down the reason for the test failure to

```
error running docker: 404 Client Error for http+docker://localhost/v1.43/images/create?tag=python3.12&fromImage=lambci%2Flambda: Not Found ("manifest for lambci/lambda:python3.12 not found: manifest unknown: manifest unknown")
```

I've pushed https://github.com/DemocracyClub/dc_logging/pull/657/commits/38c5032d66299776bfb76f948e624fc48988944c which will now log that error instead of swallowing it.

I've also raised an issue upstream over on the moto repo https://github.com/getmoto/moto/issues/8236 Lets see where that goes..

**Comment by chris48s** [comment]
TODO: once moto 5.0.18 is released, upgrade to that and work round this for now by setting `MOTO_DOCKER_LAMBDA_IMAGE=shogo82148/lambda-python:3.12.2024.10.18` when we run pytest

File: Pipfile [hunk 2]
```diff
@@ -29,12 +29,12 @@ pytest-cov = "*"
...
+moto = {extras = ["apigateway", "firehose", "awslambda", "proxy", "sts"], version = "==5.0.19.dev7"}
 
 [requires]
 python_version = "3.12"
 
 [scripts]
 ruff-check = "ruff check ."
 ruff-format-check = "ruff format --check ."
-pytest = "pytest --ruff --junitxml=test-results/junit.xml --cov=dc_logging_client"
+pytest = "sh -c \"MOTO_DOCKER_LAMBDA_IMAGE=shogo82148/lambda-python:3.12.2024.10.18 pytest --ruff --junitxml=test-results/junit.xml --cov=dc_logging_client\""
```
Comment (chris48s) [inline Pipfile:40]: This gets our tests passing by specifying a custom third-party lambda/python3.12 docker image. It is something I'd hope to be able to remove with a future release of moto.

**Review by symroe — APPROVED** [review]

**Pull request merged**